bam/reader.go

```diff
@@ -101,6 +99,52 @@ const (
 	AllVariableLengthData        // Omit sequence, quality and auxiliary data.
 )
 
+// cursor is a helper struct to keep offset into bytes as they are parsed.
+type cursor struct {
+	off int
+	raw []byte
+}
+
+func (c *cursor) bytes(n int) []byte {
+	if n < 0 {
+		n = len(c.raw) - c.off
+	}
+	s := c.off
+	c.off += n
+	return c.raw[s:c.off]
+}
+
+func (c *cursor) readUint8() uint8 {
+	c.off++
+	return c.raw[c.off-1]
+}
+
+func (c *cursor) readUint16() uint16 {
+	return binary.LittleEndian.Uint16(c.bytes(2))
+}
+
+func (c *cursor) readInt32() int32 {
+	return int32(binary.LittleEndian.Uint32(c.bytes(4)))
+}
+
+func (c *cursor) readUint32() uint32 {
+	return binary.LittleEndian.Uint32(c.bytes(4))
+}
+
+func newcursor(r *bgzf.Reader, buf [4]byte) (*cursor, error) {
+	c := &cursor{raw: buf[:4]}
+	size := int(c.readInt32())
+	c.off, c.raw = 0, make([]byte, size)
+	n, err := r.Read(c.raw)
+	if err != nil {
+		return nil, err
+	}
+	if n != size {
```

Comment:
Fleshing it out, it makes it worse. Ignore both of these (this one and [here](https://github.com/biogo/hts/pull/38/files/6fd6756ce6ffa0c469b7e0df2fa00c3cadcebaea#diff-79ef3a5e756c991412daca300726aa6a)). Sorry.